sql: support `COMMENT ON INDEX`

hueypark opened the pull request:
Informs #41316.

This patch introduces support for index comments.

The syntax to set or delete a comment is the same as postgres:
`COMMENT ON INDEX ... IS ...`. See:
https://www.postgresql.org/docs/10/static/sql-comment.html

Release note (sql change): CockroachDB now supports associating
comments to SQL index using PostgreSQL's `COMMENT ON INDEX`.
This comment can be checked with the `SHOW INDEXES FROM ... WITH COMMENT` command.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/41555)
<!-- Reviewable:end -->

jordanlewis reviewed (approved):
Very nice! Can you add a test case to `parse_test.go` analogous to the other ones you added before?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41555#-:-LrLuzwz0C_7OYmH9t3y:b-lmj2n2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @hueypark)

___
*[pkg/sql/comment_on_index.go, line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41555#-LrLu5df49iCelHDZy2c:-LrLu5df49iCelHDZy2d:bxut4yg) ([raw file](https://github.com/cockroachdb/cockroach/blob/03fcc30458e066fc5d0fa8f8bf17b6244a125d7f/pkg/sql/comment_on_index.go#L28)):*
> ```Go
> }
> 
> // CommentOnIndex add comment on a index.
> ```

Grammar: `CommentOnIndex adds a comment on an index.`
___
*[pkg/sql/pg_catalog.go, line 1166 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41555#-LrLvO7O0RDZY_573isx:-LrLvO7O0RDZY_573isy:b-apvjhj) ([raw file](https://github.com/cockroachdb/cockroach/blob/03fcc30458e066fc5d0fa8f8bf17b6244a125d7f/pkg/sql/pg_catalog.go#L1166)):*
> ```Go
> 			case keys.IndexCommentType:
> 				objID = makeOidHasher().IndexOid(
> 					sqlbase.ID(tree.MustBeDInt(comment[0])),
> ```

Seems like you should reuse your variables from above rather than `comment[n]` - it will be more clear.
___
*[pkg/sql/truncate.go, line 487 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41555#-LrLuUWeEXv97T4a1UQt:-LrLuUWeEXv97T4a1UQu:b7q9jlx) ([raw file](https://github.com/cockroachdb/cockroach/blob/03fcc30458e066fc5d0fa8f8bf17b6244a125d7f/pkg/sql/truncate.go#L487)):*
> ```Go
> }
> 
> // reassignIndexComment reassign comment on index.
> ```

`reassignIndexComment reassigns a comment on an index`
___
*[pkg/sql/sem/tree/comment_on_index.go, line 15 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41555#-LrLurSA6flop7fhRaPi:-LrLurSA6flop7fhRaPj:b-pno7mi) ([raw file](https://github.com/cockroachdb/cockroach/blob/03fcc30458e066fc5d0fa8f8bf17b6244a125d7f/pkg/sql/sem/tree/comment_on_index.go#L15)):*
> ```Go
> import "github.com/cockroachdb/cockroach/pkg/sql/lex"
> 
> // CommentOnIndex represents an COMMENT ON INDEX statement.
> ```

`represents a COMMENT ON INDEX statement`


<!-- Sent from Reviewable.io -->

hueypark committed:
sql: support `COMMENT ON INDEX`

This patch introduces support for index comments.

The syntax to set or delete a comment is the same as postgres:
`COMMENT ON INDEX ... IS ...`. See:
https://www.postgresql.org/docs/10/static/sql-comment.html

Release note (sql change): CockroachDB now supports associating
comments to SQL index using PostgreSQL's `COMMENT ON INDEX`.
This comment can be checked with the `SHOW INDEXES FROM ... WITH COMMENT` command.

hueypark reviewed (commented):
TFTR!
I added test cases to `parse_test.go`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41555#-:-LrOEZx45lphmuFR_zzI:b-vspl9n)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis)

___
*[pkg/sql/comment_on_index.go, line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41555#-LrLu5df49iCelHDZy2c:-LrOG1yVCygkRxRLHU_9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/03fcc30458e066fc5d0fa8f8bf17b6244a125d7f/pkg/sql/comment_on_index.go#L28)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Grammar: `CommentOnIndex adds a comment on an index.`
</blockquote></details>

Done.
___
*[pkg/sql/pg_catalog.go, line 1166 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41555#-LrLvO7O0RDZY_573isx:-LrOQE3n0gLKLeXZ6qxi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/03fcc30458e066fc5d0fa8f8bf17b6244a125d7f/pkg/sql/pg_catalog.go#L1166)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Seems like you should reuse your variables from above rather than `comment[n]` - it will be more clear.
</blockquote></details>

Done.
___
*[pkg/sql/truncate.go, line 487 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41555#-LrLuUWeEXv97T4a1UQt:-LrOGy_n0FID9lvjYzDq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/03fcc30458e066fc5d0fa8f8bf17b6244a125d7f/pkg/sql/truncate.go#L487)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

`reassignIndexComment reassigns a comment on an index`
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/comment_on_index.go, line 15 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41555#-LrLurSA6flop7fhRaPi:-LrOHG0C0LLDK_eEfvta:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/03fcc30458e066fc5d0fa8f8bf17b6244a125d7f/pkg/sql/sem/tree/comment_on_index.go#L15)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

`represents a COMMENT ON INDEX statement`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
Looks good! Thank you.

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41555#-:-LrW4jlH8e8YFHg_vB6i:b-6a2nvr)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->
